pkg/operator/encryption/key_controller.go

```diff
@@ -266,22 +266,27 @@ func (c *keyController) getCurrentModeAndExternalReason() (mode, string, error)
 }
 
 // TODO unit tests
-func needsNewKey(grKeys keysState, currentMode mode, externalReason string) (uint64, string, bool) {
-	// if the length of read secrets is more than one (i.e. we have more than just the write key),
-	// then we haven't successfully migrated and removed old keys so you should wait before generating more keys.
-	if len(grKeys.readSecrets) > 1 {
-		return 0, "", false
-	}
-
+func needsNewKey(grKeys keysState, currentMode mode, externalReason string, encryptedGRs []schema.GroupResource) (uint64, string, bool) {
 	// we always need to have some encryption keys unless we are turned off
 	if len(grKeys.readSecrets) == 0 {
 		return 0, "no-secrets", currentMode != identity
 	}
 
 	latestKey, latestKeyID := grKeys.latestKey()
```

Comment:
will this function behave correctly when a secret was deleted?

Reply:
yes. The desired state includes deleted keys with the right IDs. It's what you will see here, and the check below will notice that we have to create a new key in a secret.